awx/main/tasks/receptor.py

```diff
@@ -266,139 +267,119 @@ def __init__(self, task, runner_params=None):
+                    self.receptor_ctl.simple_command(f"work release {unit_id}")
                 except Exception:
-                    logger.exception(f"Error releasing work unit {self.unit_id}.")
+                    logger.exception(f"Error releasing work unit {unit_id}.")
+
+    def _run(self):
+        payload_writer, payload_reader = socket.socketpair()
+
+        # This ThreadPoolExecutor runs for the duration of the job.
+        # The cancel_func pattern is intended to guard againsnt any situation where we may
+        # end up stuck while reading or writing from the receptor socket.
+        with concurrent.futures.ThreadPoolExecutor(max_workers=3) as executor:
+            # Launch transmitter and submitter at the same time so the payload
+            # can be read by receptor while it's still being written.
+            transmitter_future = executor.submit(self.transmitter, payload_writer)
+
+            # a file-like interface for receptor to read the payload from
+            payload_reader_file = payload_reader.makefile('rb')
+            submitter_future = executor.submit(self.submitter, payload_reader_file)
+
+            def cancel_func():
+                payload_writer.shutdown(socket.SHUT_RDWR)
+                payload_writer.close()
+                payload_reader.shutdown(socket.SHUT_RDWR)
+                payload_reader.close()
+                payload_reader_file.close()
+                self.receptor_ctl._socket.shutdown(socket.SHUT_RDWR)
+                self.receptor_ctl._socket.close()
+                self.receptor_ctl._sockfile.close()
+
+            # This will return either when the job is canceled, or when the transmitter and submitter futures return.
+            cancel_watcher_future = executor.submit(self.cancel_watcher, [transmitter_future, submitter_future], cancel_func)
+            res = cancel_watcher_future.result()
+            if res and res.status == "canceled":
+                return res, None
+
+            unit_id = submitter_future.result()
+
+            # Artifacts are an output, but sometimes they are an input as well
+            # this is the case with fact cache, where clearing facts deletes a file, and this must be captured
+            self.clean_artifacts()
+
+            work_results_future = executor.submit(self.get_work_results_socket, unit_id)
+
+            def cancel_func():
+                self.receptor_ctl._socket.shutdown(socket.SHUT_RDWR)
+                self.receptor_ctl._socket.close()
+                self.receptor_ctl._sockfile.close()
+
+            cancel_watcher_future = executor.submit(self.cancel_watcher, [work_results_future], cancel_func)
+            # This will return either when the job is canceled, or when the worker_results future returns.
+            res = cancel_watcher_future.result()
+            if res and res.status == "canceled":
+                return res, unit_id
+
+            resultsock, resultfile = work_results_future.result()
```

Comment:
All `work_results` does is return the socket. It seems somewhat extraordinary that _it_ would be hanging.

https://github.com/ansible/receptor/blob/ba3ed4532509e9d92a4d0bd89c0a284ee13f58a0/receptorctl/receptorctl/socket_interface.py#L248

It does read from the general communication socket to receptor, so it absolutely could hang.